sql/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/streaming/TransformWithStateConnectSuite.scala

```diff
@@ -0,0 +1,489 @@
+      key: String,
+      inputRows: Iterator[(String, String)],
+      timerValues: TimerValues): Iterator[(String, String)] = {
+    var count = 0
+    var ttlCount = 0
+    var ttlListStateCount = 0
+    var ttlMapStateCount = 0
+    if (countState.exists()) {
+      count = countState.get()
+    }
+    if (ttlCountState.exists()) {
+      ttlCount = ttlCountState.get()
+    }
+    if (ttlListState.exists()) {
+      for (value <- ttlListState.get()) {
+        ttlListStateCount += value
+      }
+    }
+    if (ttlMapState.exists()) {
+      ttlMapStateCount = ttlMapState.getValue(key)
+    }
+    for (_ <- inputRows) {
+      count += 1
+      ttlCount += 1
+      ttlListStateCount += 1
+      ttlMapStateCount += 1
+    }
+    countState.update(count)
+    if (key != "0") {
+      ttlCountState.update(ttlCount)
+      ttlListState.put(Array(ttlListStateCount, ttlListStateCount))
+      ttlMapState.updateValue(key, ttlMapStateCount)
+    }
+    val output = List(
+      (s"count-$key", count.toString),
+      (s"ttlCount-$key", ttlCount.toString),
+      (s"ttlListState-$key", ttlListStateCount.toString),
+      (s"ttlMapState-$key", ttlMapStateCount.toString))
+    output.iterator
+  }
+}
+
+class TransformWithStateConnectSuite extends QueryTest with RemoteSparkSession with Logging {
+  val testData: Seq[(String, String)] = Seq(("a", "1"), ("b", "1"), ("a", "2"))
+  val twsAdditionalSQLConf = Seq(
+    "spark.sql.streaming.stateStore.providerClass" ->
+      "org.apache.spark.sql.execution.streaming.state.RocksDBStateStoreProvider",
+    "spark.sql.shuffle.partitions" -> "5",
+    "spark.sql.session.timeZone" -> "UTC",
+    "spark.sql.streaming.noDataMicroBatches.enabled" -> "false")
+
+  test("transformWithState - streaming with state variable, case class type") {
+    withSQLConf(twsAdditionalSQLConf: _*) {
+      val session: SparkSession = spark
+      import session.implicits._
+
+      spark.sql("DROP TABLE IF EXISTS my_sink")
+
+      withTempPath { dir =>
+        val path = dir.getCanonicalPath
+        testData
+          .toDS()
+          .toDF("key", "value")
+          .repartition(3)
+          .write
+          .parquet(path)
+
+        val testSchema =
+          StructType(Array(StructField("key", StringType), StructField("value", StringType)))
+
+        val q = spark.readStream
+          .schema(testSchema)
+          .option("maxFilesPerTrigger", 1)
+          .parquet(path)
+          .as[InputRowForConnectTest]
+          .groupByKey(x => x.key)
+          .transformWithState[OutputRowForConnectTest](
+            new BasicCountStatefulProcessor(),
+            TimeMode.None(),
+            OutputMode.Update())
+          .writeStream
+          .format("memory")
+          .queryName("my_sink")
+          .start()
+
+        try {
+          q.processAllAvailable()
+          eventually(timeout(30.seconds)) {
+            checkDataset(
+              spark.table("my_sink").toDF().as[(String, String)].orderBy("key"),
```

Comment:
If you don't put isSorted = true, checkDataset will sort each of parameter again with string representation of the Row. See QueryTest.prepareAnswer().

That said, orderBy here is effectively no-op and you don't actually need it.